use for_each instead of count to create folders

adrian-gierakowski opened the pull request:
This prevents renaming of folders when content/order of the folder names list
changes.

Addresses issue mentioned in: https://github.com/terraform-google-modules/cloud-foundation-fabric/tree/73b7613cc062491305ebfabf8a8b1db0c53fc55b/foundations#things-to-be-aware-of

WARNING: this is a breaking change as the terraform state created before this
change will be incompatible with state expected after the change. This can
be fixed with `terraform state mv`. For example, given the code from the
README.md of this repo, the state would have to be migrated with the following
commands:

```
terraform state mv module.folders.google_folder.folders[0] 'module.folders.google_folder.folders["dev"]'
terraform state mv module.folders.google_folder.folders[1] 'module.folders.google_folder.folders["staging"]'
terraform state mv module.folders.google_folder.folders[2] 'module.folders.google_folder.folders["production"]'
```

ludoo commented on main.tf:
```diff
@@ -15,16 +15,21 @@
 resource "google_folder" "folders" {
-  count        = length(var.names)
-  display_name = "${local.prefix}${element(var.names, count.index)}"
+  for_each = local.names_map
```
You can do `for_each` on lists, no need to create a map on `var.names`.

adrian-gierakowski replied:
thanks! fixed in: 332affb

ludoo commented on main.tf:
```diff
@@ -15,16 +15,21 @@
   prefix = var.prefix == "" ? "" : "${var.prefix}-"
+  names_map = zipmap(var.names, var.names)
+  folders = values(google_folder.folders)
+  first_folder = local.folders[0]
```
This should be the folder corresponding to the first name in `var.names`, not the first folder in lexical order which is what you're setting here.

adrian-gierakowski replied:
fixed in 5fe55f9

ludoo commented on main.tf:
```diff
@@ -15,16 +15,21 @@
 locals {
   prefix = var.prefix == "" ? "" : "${var.prefix}-"
+  names_map = zipmap(var.names, var.names)
+  folders = values(google_folder.folders)
```
This should respect the order of the `var.names` variable, not be in lexical order.

adrian-gierakowski replied:
fixed in 5fe55f9

adrian-gierakowski commented:
@ludoo thanks for review. I've addressed your comments and also improved creation of google_folder_iam_binding. Unfortunately tests are failing as I'm not sure how to run them locally and was only guessing what should change to make them pass

adrian-gierakowski commented:
btw. I've created a related pr on [terraform-google-cloud-storage](https://github.com/terraform-google-modules/terraform-google-cloud-storage/pull/24)

ludoo commented:
Hey Adrian, thanks for this.

To run tests locally, check out the [CONTRIBUTING.md](https://github.com/terraform-google-modules/terraform-google-folders/blob/master/CONTRIBUTING.md) doc, which explains how ti run tests using the dev container image.

> btw. I've created a related pr on [terraform-google-cloud-storage](https://github.com/terraform-google-modules/terraform-google-cloud-storage/pull/24)

There's already a PR there I sent a while ago, can you check it out and see if it addresses your use case? There was a bit of discussion around the interface, but apart from that it should be ready to go.

morgante commented:
FYI tests will likely fail until we merge #23.

adrian-gierakowski commented:
@morgante cool, thanks for the info. The test pass for me locally now. I could try rebasing on top of #23 now to see if they pass on CI

adrian-gierakowski commented:
one of the jobs still failing after rebase
not sure why since I can't see the logs

morgante commented:
@adrian-gierakowski It's the linter that's failing, try running `make docker_test_lint`.

adrian-gierakowski commented:
all green

adrian-gierakowski commented:
Sure, I’m travelling right now, should be able to do this by the end of the
week.


On Tue, 3 Dec 2019 at 09:34, Ludovico Magnocavallo <notifications@github.com>
wrote:

> *@ludoo* approved this pull request.
>
> Thanks! Could you add a migration doc and update CHANGELOG? You can
> probably use this as an example
>
>
> https://github.com/terraform-google-modules/terraform-google-cloud-storage/blob/ludo-fix-issue-18/docs/upgrading_to_v2.md
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/terraform-google-modules/terraform-google-folders/pull/21?email_source=notifications&email_token=AACQTQMZBPV4DLTCVBB7XHTQW2KDTA5CNFSM4JTU7ZT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCNZU3YQ#pullrequestreview-326323682>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACQTQOX6EVUPVWYSJVYEJTQW2KDTANCNFSM4JTU7ZTQ>
> .
>
-- 
Adrian Gierakowski

adrian-gierakowski commented:
@ludoo I'm sorry but I will not be able to take care of the migration script and docs this week as I have a project to wrap up. Maybe this could be merge as is and a seperate issue created for the script/docs?

adrian-gierakowski commented:
> Hey Adrian, thanks for this.
> 
> To run tests locally, check out the [CONTRIBUTING.md](https://github.com/terraform-google-modules/terraform-google-folders/blob/master/CONTRIBUTING.md) doc, which explains how ti run tests using the dev container image.
> 
> > btw. I've created a related pr on [terraform-google-cloud-storage](https://github.com/terraform-google-modules/terraform-google-cloud-storage/pull/24)
> 
> There's already a PR there I sent a while ago, can you check it out and see if it addresses your use case? There was a bit of discussion around the interface, but apart from that it should be ready to go.

yes that PR (https://github.com/terraform-google-modules/terraform-google-cloud-storage/pull/22) work for me, any indication when it could be merged?

adrian-gierakowski commented:
@morgante any chance of getting this merged? I just squashed the commits and rebased it on master.

adrian-gierakowski commented:
integration tests are failing however I can't see the logs, and trying to run them locally gives me the following error: 

```
Error: Provider configuration not present

To work with
module.folders-project.module.project-factory.null_resource.delete_default_compute_service_account[0]
its original provider configuration at
provider["registry.terraform.io/-/null"] is required, but it has been removed.
This occurs when a provider configuration is removed while objects created by
that provider still exist in the state. Re-add the provider configuration to
destroy
module.folders-project.module.project-factory.null_resource.delete_default_compute_service_account[0],
after which you can remove the provider configuration again.
```

just to make sure I'm doing it right, I've set the env vars as described [here](https://github.com/terraform-google-modules/terraform-google-folders/blob/master/CONTRIBUTING.md#test-environment) and ran: `make docker_test_prepare` from the root of the repo.

adrian-gierakowski commented:
this doesn't seem right: 
https://github.com/terraform-google-modules/terraform-google-folders/blob/ac913087cb44d049c7d62466aa399af60bd9851b/Makefile#L21

looks like it should be `0.12.0`:
https://github.com/terraform-google-modules/terraform-google-folders/blob/ac913087cb44d049c7d62466aa399af60bd9851b/build/int.cloudbuild.yaml#L41

morgante commented:
Yeah we should pin to 0.12.0.

adrian-gierakowski commented:
@morgante after updating docker image to 0.12.0 I manage to run the tests. I got a bunch of errors which look like this:
```
Error applying IAM policy for folder "folders/184570595698": Error setting IAM policy for folder "folders/184570595698": googleapi: Error 400: Group gcp-ddt-qa@apszaz.com does not exist., badRequest
```

I fixed it by changing `gcp-ddt-qa@apszaz.com ` etc. to some users which actually exists on my GCP org and both `make docker_test_lint` and `make docker_test_integration` succeed on my side now. Not sure why the cloudbuild jobs are failing. Could you please have a look at it? Thanks!

bharathkkb commented on test/fixtures/simple_example/variables.tf:
```diff
@@ -36,13 +36,13 @@ variable "names" {
-    "group:test-gcp-ops@test.infra.cft.tips",
-  ]
+  default = {
+    dev        = "group:gcp-ddt-developers@apszaz.com",
```
We should keep the older names which exist for testing in our org.

adrian-gierakowski replied:
fixed in 3b439120a7ccb1c9de02ad0b099a653ccf73c493

bharathkkb reviewed (commented):
Thanks for the PR!
Tests seem to be failing because we need to bump the provider [version]( https://github.com/terraform-google-modules/terraform-google-folders/blob/ac913087cb44d049c7d62466aa399af60bd9851b/test/setup/versions.tf#L22) in setup as we released a new PF version which has features from the `3.38`

bharathkkb commented:
@adrian-gierakowski could you also update the setup provider like mentioned in https://github.com/terraform-google-modules/terraform-google-folders/pull/21#pullrequestreview-497030409

adrian-gierakowski commented:
> @adrian-gierakowski could you also update the setup provider like mentioned in [#21 (review)](https://github.com/terraform-google-modules/terraform-google-folders/pull/21#pullrequestreview-497030409)

done in 24c0afd, all tests passing now, thanks!

morgante reviewed (approved):
Thanks for the PR, we'll also need an upgrade guide before release.
